common/src/main/scala/uk/ac/wellcome/utils/TryBackoff.scala

```diff
@@ -19,30 +59,61 @@ trait TryBackoff extends Logging {
       f()
     } match {
       case Success(_) => 0
-      case Failure(_) =>
-        error(s"Failed to run (attempt: $attempt)")
+      case Failure(e) =>
+        error(s"Failed to run (attempt: $attempt)", e)
         attempt + 1
     }
 
-    if (numberOfAttempts > maxAttempts)
+    // If we're in non-continuous mode and the last attempt succeeded,
+    // we should return immediately.
+    if (numberOfAttempts == 0 && !continuous) {
+      return
```

Comment:
Sorry, last comment: this is also not very functional...We should avoid returning in the middle of a function if possible. Try to use a value instead maybe? Something like:
```val shouldReSchedule = numberOfAttempts != 0 || continuous```
And then down below you use it to decide whether to schedule the function or not like so:
```
if(shouldReSchedule) {
      val cancellable = system.scheduler.scheduleOnce(waitTime milliseconds)(
        run(f, system, attempt = numberOfAttempts))
      maybeCancellable = Some(cancellable)
    }
```